readme.md
# Algolia Index Update

A command-line utility for web sites using Algolia (for site indexing and search) to merge an updated index with an Algolia project data source. The module is based on the process outlined in [Adding Algolia Search to Eleventy and Netlify](https://www.raymondcamden.com/2020/06/24/adding-algolia-search-to-eleventy-and-netlify) and uses the [Algolia-indexing](https://github.com/pixelastic/algolia-indexing) module as described in the article. 

In the article, @cfjedimaster shows how to use generate the Algolia index from an [Eleventy](https://www.11ty.dev/) site and use [Netlify](https://www.netlify.com/) Functions to merge the updated index and upload it to Algolia during a Netlify build process. In the article he also mentioned a 10 minute limit for executing Cloud Functions in Netlify, and that scared me a little bit. Since I'm comfortable using a local generation and deployment process for my site's Algolia index, I decided to build this command-line module to handle the task for me.

![The module in action](https://github.com/johnwargo/algolia-index-update/blob/main/images/image-02.png)

Here's a link to the NPM project:

**Note:** You don't have to use Eleventy to use this module, all it cares about is local access to an Algolia site index and the Algolia project and account API keys needed to merge the local index with the Algolia project data source.

To use the module,  you must have:

* An Algolia Application project and data source  defined in Algolia. 
* The Algolia Application ID for your site's Algolia project.
* The Admin API keys for your Algolia account.
* The name of the Algolia project's index.

## Installation 

To install the module as a dependency in the application project, open a terminal window or command prompt and execute the following command:

```shell
npm install algolia-index-update --save-dev 
```

This adds an `algolia-idxup` command you can execute through the standard `npm run` build process at the command-line or in the project's `package.json` file.

To install the module globally, open a terminal window or command prompt and execute the following command:

```shell
npm install algolia-index-update -g
```

This adds an `algolia-idxup` command you can execute from any terminal window or command prompt on the system. 

## Usage

To use the module, you must first define the environment variables listed in the following table:

| Environment Variable Name | Algolia Project Key or Value   |
| ------------------------- | ------------------------------ |
| `ALGOLIA_API_KEY`         | Algolia account Admin API key  |
| `ALGOLIA_APP_ID`          | Algolia project Application ID |
| `ALGOLIA_IDX_NAME`        | Algolia project Index name     |

The module reads these environment variables and uses their values when merging the updated index with the Algolia cloud project. It uses environment variables to enable hiding the values from the project's source code in GitHub (or some other source code repository) and allows the module to run in cloud hosting environments like Netlify.

If you use Algolia for multiple sites, you need a way to distinguish the app id and index names, so you can create a second set of credentials and append a prefix to the environment variables. For example, for my Random Errors site, I use a prefix of `RE_` and I setup my variable names like this:

| Environment Variable Name | Algolia Project Key or Value   |
| ------------------------- | ------------------------------ |
| `RE_ALGOLIA_API_KEY`      | Algolia account Admin API key  |
| `RE_ALGOLIA_APP_ID`       | Algolia project Application ID |
| `RE_ALGOLIA_IDX_NAME`     | Algolia project Index name     |

There are two options for executing the module:

```shell
algolia-idxup <path_to_the_algolia_index>
```

or

```shell
algolia-idxup <path_to_the_algolia_index> [environment_variable_prefix]
```

Lets see some examples:

To process the Algolia index at `_site/_data` with no prefix on the environment variable names, use the following:

```shell
algolia-idxup _site/algolia.json
```

To use the prefixed environment variables described in the second table, use the following

```shell
algolia-idxup _site/algolia.json RE_
```

Alternatively, when you have more than one Algolia project, you can provide the Algolia credentials via a JSON file using the `-f` or `--file` flag coupled with the path to the JSON file:

```shell
algolia-idxup _site/algolia.json RE_ -f ..\algolia-credentials.json
```

The `algolia-credentials.json` file looks like this:

```json
{
  "ALGOLIA_APP_ID": "Root_Algolia_App_ID",
  "ALGOLIA_API_KEY": "Root_Algolia_API_Key",
  "ALGOLIA_IDX_NAME": "Root_Algolia_Idx_name",
  "RE_ALGOLIA_APP_ID": "Random_Errors_App_ID",
  "RE_ALGOLIA_API_KEY": "Random =_Errors_API_Key",
  "RE_ALGOLIA_IDX_NAME": "Random_Errors_Index_Name",
}
```

Replacing, of course, the properties with the correct values for your different Algolia projects. 

The module validates the required environment variables, then confirms that it can locate the index file. With those two components in place, the module clicks and whirs for a while as it completes the index update/merge as shown below.

```text
D:\dev\11ty\random-errors-static>algolia-idxup --default
┌──────────────────────────┐
│                          │
│   Algolia Index Update   │
│                          │
└──────────────────────────┘
Index: : D:\dev\node\algolia-index-update\_site\algolia.json
Processing index for 17 articles.
√ Copying random-errors-main to random-errors-main_tmp
√ Updating settings of random-errors-main_tmp
√ Getting all records from random-errors-main_tmp
√ Moving random-errors-main_tmp to random-errors-main
√ Configuring replicas of random-errors-main
Processing completed in 43.687 seconds

D:\dev\node\algolia-index-update>
```

If you want to see more output as the command runs, add a -d to the command line:

```shell
algolia-idxup _site/algolia.json -d
```

That's pretty much it. Enjoy!

***

You can find information on many different topics on my [personal blog](http://www.johnwargo.com). Learn about all of my publications at [John Wargo Books](http://www.johnwargobooks.com).

If you find this code useful and feel like thanking me for providing it, please consider <a *href*="https://www.buymeacoffee.com/johnwargo" *target*="_blank">Buying Me a Coffee</a>, or making a purchase from [my Amazon Wish List](https://amzn.com/w/1WI6AAUKPT5P9).


changelog.md
# Changelog

## 20230522

Added option for file-based repository for Algolia credentials instead of registry-based credentials.
## 20230502 - Version 0.2.0

For my own personal use, I need to upload indices for multiple sites, so I updated the program to accept an environment variable prefix to use when collection the API key and such. 

I also removed the --default flag requiring you to provide the path to the index file manually. Previously it defaulted to `_site/algolia.json`.

algolia-index-update.js
#!/usr/bin/env node
var HighlightType;
(function (HighlightType) {
    HighlightType[HighlightType["Red"] = 0] = "Red";
    HighlightType[HighlightType["Yellow"] = 1] = "Yellow";
    HighlightType[HighlightType["Green"] = 2] = "Green";
})(HighlightType || (HighlightType = {}));
import fs from 'fs';
import path from 'path';
import { Command } from 'commander';
import indexing from 'algolia-indexing';
import boxen from 'boxen';
import chalk from 'chalk';
import dayjs from 'dayjs';
const red = HighlightType.Red;
const yellow = HighlightType.Yellow;
const green = HighlightType.Green;
const appName = 'Algolia Index Update';
const newline = "\n";
var algoliaAppIdKeyStr = 'ALGOLIA_APP_ID';
var algoliaApiKeyStr = 'ALGOLIA_API_KEY';
var algoliaIndexNameStr = 'ALGOLIA_IDX_NAME';
var inputFilePath;
function commaize(value) {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}
function writeConsole(color, highlightText, msg) {
    if (color == HighlightType.Red)
        console.log(newline + chalk.red(`${highlightText}: `) + msg + newline);
    if (color == HighlightType.Yellow)
        console.log(chalk.yellow(`${highlightText}: `) + msg);
    if (color == HighlightType.Green)
        console.log(chalk.green(`${highlightText}: `) + msg);
}
console.log(boxen(appName, { padding: 1 }));
const program = new Command();
program
    .argument('<sourcePath>', 'Relative path for the source index file')
    .argument('[variablePrefix]', 'Environment variable prefix', '')
    .option('-d, --debug', 'Debug mode')
    .option('-f, --file <filePath>', 'Enable file-based credentials (with path to file)')
    .action(async (sourcePath, variablePrefix) => {
    console.log();
    const options = program.opts();
    const debugMode = options.debug;
    if (debugMode) {
        console.log('Debug mode enabled.');
        writeConsole(yellow, 'Source Path', sourcePath);
        if (variablePrefix != '')
            writeConsole(yellow, 'Variable prefix', variablePrefix);
        if (options.file) {
            writeConsole(yellow, 'File path', options.file);
        }
    }
    inputFilePath = path.join(process.cwd(), sourcePath);
    writeConsole(yellow, 'Index path', inputFilePath);
    if (variablePrefix != '') {
        algoliaApiKeyStr = variablePrefix + algoliaApiKeyStr;
        algoliaAppIdKeyStr = variablePrefix + algoliaAppIdKeyStr;
        algoliaIndexNameStr = variablePrefix + algoliaIndexNameStr;
    }
    else {
        if (debugMode)
            writeConsole(yellow, 'Info', 'Using root environment variables.');
    }
    const algoliaCreds = { appId: '', apiKey: '', indexName: '' };
    if (options.file) {
        writeConsole(yellow, 'Info', 'Using file-based credentials.');
        let filePath = path.join(process.cwd(), options.file);
        if (!fs.existsSync(filePath)) {
            writeConsole(red, 'Error', 'Credentials file does not exist, please try again');
            process.exit(1);
        }
        let rawData = fs.readFileSync(filePath);
        let fileData = JSON.parse(rawData.toString());
        algoliaCreds.appId = fileData[algoliaAppIdKeyStr];
        algoliaCreds.apiKey = fileData[algoliaApiKeyStr];
        algoliaCreds.indexName = fileData[algoliaIndexNameStr];
    }
    else {
        writeConsole(yellow, 'Info', 'Using environment-based credentials.');
        algoliaCreds.appId = process.env[algoliaAppIdKeyStr];
        algoliaCreds.apiKey = process.env[algoliaApiKeyStr];
        algoliaCreds.indexName = process.env[algoliaIndexNameStr];
    }
    ;
    if (debugMode)
        console.dir(algoliaCreds);
    let validConfig = true;
    if (algoliaCreds.appId == undefined)
        validConfig = false;
    if (algoliaCreds.apiKey == undefined)
        validConfig = false;
    if (algoliaCreds.indexName == undefined)
        validConfig = false;
    if (!validConfig) {
        writeConsole(red, 'Error', 'One or more Algolia credentials environment variables missing.');
        process.exit(1);
    }
    try {
        if (!fs.existsSync(inputFilePath)) {
            writeConsole(red, 'Error', 'Algolia index file does not exist, please try again');
            process.exit(1);
        }
    }
    catch (err) {
        writeConsole(red, 'Error', err.message);
        process.exit(1);
    }
    let rawData = fs.readFileSync(inputFilePath);
    let idxData = JSON.parse(rawData.toString());
    console.log(`Processing index for ${idxData.length} articles.`);
    let startTime = dayjs(Date.now());
    indexing.verbose();
    try {
        await indexing.fullAtomic(algoliaCreds, idxData, {});
    }
    catch (err) {
        writeConsole(red, 'Error', err.message);
        process.exit(1);
    }
    ;
    let diffStr = commaize(Math.abs(startTime.diff(dayjs(Date.now()), 'second', true)));
    console.log(`Processing completed in ${diffStr} seconds`);
});
program.parse();


src/algolia-index-update.ts
#!/usr/bin/env node

// https://www.raymondcamden.com/2020/06/24/adding-algolia-search-to-eleventy-and-netlify

type AlgoliaCredentials = {
    appId: string;
    apiKey: string;
    indexName: string;
}

enum HighlightType {
    Red, Yellow, Green
}

// node modules
import fs from 'fs';
import path from 'path';

// Third-party modules
import { Command } from 'commander';
//@ts-ignore
import indexing from 'algolia-indexing';
import boxen from 'boxen';
import chalk from 'chalk';
import dayjs from 'dayjs'

// local constants
const red = HighlightType.Red;
const yellow = HighlightType.Yellow;
const green = HighlightType.Green;

const appName = 'Algolia Index Update';
const newline = "\n";

var algoliaAppIdKeyStr = 'ALGOLIA_APP_ID';
var algoliaApiKeyStr = 'ALGOLIA_API_KEY';
var algoliaIndexNameStr = 'ALGOLIA_IDX_NAME';

var inputFilePath;

function commaize(value: number) {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

function writeConsole(color: HighlightType, highlightText: string, msg: string) {
    if (color == HighlightType.Red) console.log(newline + chalk.red(`${highlightText}: `) + msg + newline);
    if (color == HighlightType.Yellow) console.log(chalk.yellow(`${highlightText}: `) + msg);
    if (color == HighlightType.Green) console.log(chalk.green(`${highlightText}: `) + msg);
}

console.log(boxen(appName, { padding: 1 }));
const program = new Command();
program
    .argument('<sourcePath>', 'Relative path for the source index file')
    .argument('[variablePrefix]', 'Environment variable prefix', '')
    .option('-d, --debug', 'Debug mode')
    .option('-f, --file <filePath>', 'Enable file-based credentials (with path to file)')
    .action(async (sourcePath, variablePrefix) => {
        console.log();

        const options = program.opts();
        const debugMode = options.debug;
        if (debugMode) {
            console.log('Debug mode enabled.');
            writeConsole(yellow, 'Source Path', sourcePath);
            if (variablePrefix != '') writeConsole(yellow, 'Variable prefix', variablePrefix);
            if (options.file) {
                writeConsole(yellow, 'File path', options.file);
            }
        }

        inputFilePath = path.join(process.cwd(), sourcePath);
        writeConsole(yellow, 'Index path', inputFilePath);

        if (variablePrefix != '') {
            algoliaApiKeyStr = variablePrefix + algoliaApiKeyStr;
            algoliaAppIdKeyStr = variablePrefix + algoliaAppIdKeyStr;
            algoliaIndexNameStr = variablePrefix + algoliaIndexNameStr;
        } else {
            if (debugMode) writeConsole(yellow, 'Info', 'Using root environment variables.');
        }

        const algoliaCreds: AlgoliaCredentials = { appId: '', apiKey: '', indexName: '' }

        if (options.file) {
            writeConsole(yellow, 'Info', 'Using file-based credentials.');
            let filePath = path.join(process.cwd(), options.file);
            if (!fs.existsSync(filePath)) {
                writeConsole(red, 'Error', 'Credentials file does not exist, please try again');
                process.exit(1);
            }
            let rawData = fs.readFileSync(filePath);
            let fileData = JSON.parse(rawData.toString());
            // grab the credentials from the file
            algoliaCreds.appId = fileData[algoliaAppIdKeyStr];
            algoliaCreds.apiKey = fileData[algoliaApiKeyStr];
            algoliaCreds.indexName = fileData[algoliaIndexNameStr];
        } else {
            writeConsole(yellow, 'Info', 'Using environment-based credentials.');
            // grab the credentials from the environment
            algoliaCreds.appId = process.env[algoliaAppIdKeyStr]!;
            algoliaCreds.apiKey = process.env[algoliaApiKeyStr]!;
            algoliaCreds.indexName = process.env[algoliaIndexNameStr]!;
        };

        // TODO: reformat algoliaCreds into something you can output using writeConsole (json.stringify)
        if (debugMode) console.dir(algoliaCreds);

        // Validate the Algolia credentials
        let validConfig = true;
        if (algoliaCreds.appId == undefined) validConfig = false;
        if (algoliaCreds.apiKey == undefined) validConfig = false;
        if (algoliaCreds.indexName == undefined) validConfig = false;
        if (!validConfig) {
            writeConsole(red, 'Error', 'One or more Algolia credentials environment variables missing.');
            process.exit(1);
        }

        try {
            if (!fs.existsSync(inputFilePath)) {
                writeConsole(red, 'Error',
                    'Algolia index file does not exist, please try again');
                process.exit(1);
            }
        } catch (err: any) {
            writeConsole(red, 'Error', err.message);
            process.exit(1);
        }

        let rawData = fs.readFileSync(inputFilePath);
        let idxData = JSON.parse(rawData.toString());

        console.log(`Processing index for ${idxData.length} articles.`);
        let startTime = dayjs(Date.now());
        indexing.verbose();
        try {
            await indexing.fullAtomic(algoliaCreds, idxData, {});
        } catch (err: any) {
            writeConsole(red, 'Error', err.message);
            process.exit(1);
        };
        // let endTime = dayjs(Date.now());
        let diffStr = commaize(Math.abs(startTime.diff(dayjs(Date.now()), 'second', true)));
        console.log(`Processing completed in ${diffStr} seconds`);
    });

program.parse();